### PR TITLE
Error on invalid LAMMPS instance

### PR DESCRIPTION
**Pull request by Joroks**
I've taken a few things from #34 and made sure that we check the lammps instance before using it for all of our public API

### CODE REVIEW

**Comment by vchuravy** [comment]
Ah interesting, so the issue with checking before an operation is that the error is "far" away from the original location.
Once should always check after so that the error is thrown close to the offending operation. 
It's probably sensible to check before as well to make sure we don't pass an invalid state.

**Comment by Joroks** [comment]
The main issue I'm adressing here is that it's not valid to pass null pointers as the handle for a lammps instance (with very few exceptions). Doing this currently causes a crash. I've decided to include this check in the `check` method for simplicity. We could also seperate that out though.

**Comment by vchuravy** [comment]
Actually we can define next to our `unsafe_convert` function:

```
Base.cconvert(::Type{Ptr{Cvoid}}, lmp::LMP) = lmp.lmp == C_NULL ? error("") : lmp
```

So that whenever a LMP instance is passed to a C-API we emit an error check.

File: src/LAMMPS.jl [hunk 1]
```diff
@@ -139,14 +139,18 @@ mutable struct LMP
...
         return this
     end
 end
-Base.unsafe_convert(::Type{Ptr{Cvoid}}, lmp::LMP) = lmp.handle
+
+function Base.unsafe_convert(::Type{Ptr{Cvoid}}, lmp::LMP)    
+    lmp.handle == C_NULL && error("The LMP object doesn't point to a valid LAMMPS instance! "
+            * "This is usually caused by calling `LAMMPS.close!` or through serialization and deserialization.")
+    return lmp.handle
+end
```
Comment (vchuravy) [inline src/LAMMPS.jl:153]:
```suggestion
function Base.cconvert(::Type{Ptr{Cvoid}}, lmp::LMP)    
    lmp.handle == C_NULL && error("The LMP object doesn't point to a valid LAMMPS instance! "
            * "This is usually caused by calling `LAMMPS.close!` or through serialization and deserialization.")
    return lmp
end
Base.unsafe_convert(::Type{Ptr{Cvoid}}, lmp::LMP) = lmp.handle
```

File: src/LAMMPS.jl [hunk 2]
```diff
@@ -191,14 +194,11 @@ function Base.showerror(io::IO, err::LAMMPSError)
     print(io, "LAMMPSError: ", err.msg)
 end
 
-function check(lmp::LMP)
+function check(lmp::Union{LMP, Ptr{Cvoid}})
```
Comment (vchuravy) [inline src/LAMMPS.jl:197]: Why the change to allow `void*`

Comment (vchuravy) [reply src/LAMMPS.jl:197]: Same for LAMMPSError

File: src/LAMMPS.jl [hunk 3]
```diff
@@ -139,13 +139,18 @@ mutable struct LMP
                 handle = API.lammps_open_no_mpi(length(args), args, C_NULL)
             end
         end
-
+        check(handle)
         this = new(handle)
```
Comment (vchuravy) [inline src/LAMMPS.jl:143]:
```suggestion
        this = new(handle)
        check(this)
```

File: src/LAMMPS.jl (hunk 2 above)
Comment (Joroks) [reply src/LAMMPS.jl:197]: Look at the [Documentation](https://docs.lammps.org/Library_utility.html#_CPPv416lammps_has_errorPv) for `lammps_has_error` and `lammps_get_last_error_message`. They specifically allow the handle to be NULL in case we create a lammps instance with invalid command line arguments. However, because we check that our handle isn't Null before converting our LMP object to the Pointer, we'd get a confusing error message.

File: src/LAMMPS.jl (hunk 3 above)
Comment (Joroks) [reply src/LAMMPS.jl:143]: doing this to allow a NULL pointer to be passed to check.

File: src/LAMMPS.jl (hunk 1 above)
Comment (Joroks) [reply src/LAMMPS.jl:153]: I'm not quite sure what difference that would make. Could you explain?

Comment (vchuravy) [reply src/LAMMPS.jl:153]: The expectation in Julia is that `unsafe_convert` is well `unsafe` it shouldn't perform error checking.
Most notably the invariant that Julia guarantees is that the input argument to `unsafe_convert` remains rooted during the duration of the `ccall`, but the output is not rooted.

For argument conversion (or in our case argument checking) one can use [`cconvert`](https://docs.julialang.org/en/v1/base/c/#Base.cconvert)


```
ccall(:f, Cvoid, (Ptr{Cvoid},), x)
```

turns into:
```
obj = Base.cconvert(Ptr{Cvoid}, x)
GC.@preserve obj begin
     raw_obj = Base.unsafe_convert(Ptr{Cvoid},obj)
     ccall(:f, Cvoid, (Ptr{Cvoid},), raw_obj)
end
```
It can sometimes matter how much code is inside the `GC.@preserve`.

File: src/LAMMPS.jl (hunk 2 above)
Comment (vchuravy) [reply src/LAMMPS.jl:197]: I would then rather inline the error handling code. This ought to be the only case where we accept a raw pointer in the Julia API.

Comment (Joroks) [reply src/LAMMPS.jl:197]: My reasoning was that `check` was mostly for internal use anyways. I'll go ahead and inline the error handling then.

**Review by vchuravy — APPROVED** [review]
